File: Java/Searching/BinarySearch.java
/* 
Explanation:
- The binarySearch function takes an array `arr` and a target value `target` as parameters.
- It initializes `left` and `right` variables to keep track of the search range within the array.
- The while loop continues until the `left` pointer is less than or equal to the `right` pointer.
- Inside the loop, it calculates the `mid` index using `(left + right) / 2` to divide the search range into two halves.
- If the target value is found at the `mid` index, it returns the index.
- If the target value is greater than the value at the `mid` index, it updates the `left` pointer to `mid + 1` to search in the right half.
- If the target value is smaller than the value at the `mid` index, it updates the `right` pointer to `mid - 1` to search in the left half.
- If the loop is completed without finding the target, it returns -1.
- In the main method, we create an example array `arr` and a target value `target`.
- We call the `binarySearch` function with these parameters and store the returned index in the `index` variable.
- Finally, we check if the index is not -1 and print the appropriate message.

Time Complexity: O(log n) - Binary search runs in logarithmic time, as it halves the search range with each step.

Space Complexity: O(1) - Binary search has a constant space complexity as it does not require any additional space that grows with the input size.

Sample Input:
arr = [1, 2, 3, 4, 5, 6, 7]
target = 4

Sample Output:
Element found at index 3

*/

//Here's the code for Binary Search in Java:

public class BinarySearch {
    // Binary Search function
    public static int binarySearch(int[] arr, int target) {
        int left = 0;
        int right = arr.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            // Check if target is present at the mid index
            if (arr[mid] == target) {
                return mid;
            }

            // If target is greater, ignore the left half
            if (arr[mid] < target) {
                left = mid + 1;
            }
            // If target is smaller, ignore the right half
            else {
                right = mid - 1;
            }
        }

        //  Return -1 if the target is not found
        return -1;
    }

    // Main method to test the code
    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 4, 5, 6, 7};
        int target = 4;
        int index = binarySearch(arr, target);

        if (index != -1) {
            System.out.println("Element found at index " + index);
        } else {
            System.out.println("Element not found");
        }
    }
}

File: C++/Searching/BinarySearch.cpp
/*
Explanation:
- The `binarySearch` function takes a sorted array, its size, and a key as input. It performs binary search, which works by repeatedly dividing the array in half and comparing the middle element (mid) with the key.
- We initialize two pointers `low` and `high` to mark the range of the search. Initially, `low` points to the first element of the array, and `high` points to the last element.
- In each iteration of the `while` loop, we calculate the middle index (`mid`) using `(low + high) / 2` to avoid integer overflow. If the element at the middle index is equal to the key, we return the mid index.
- If the element at `mid` is less than the key, it means the key lies in the right half of the array, so we update `low` to `mid + 1` to search the right half.
- If the element at `mid` is greater than the key, it means the key lies in the left half of the array, so we update `high` to `mid - 1` to search the left half.
- We repeat this process until the `low` and `high` pointers cross each other. If the key is not found, we return -1.

Time Complexity: The time complexity of binary search is O(log N), where N is the size of the array. This is because in each iteration, we reduce the search space by half, which results in a much faster search compared to linear search for large arrays.

Space Complexity: The space complexity is O(1) since the algorithm only uses a constant amount of extra memory for the `low`, `high`, `mid`, `size`, and `key` variables, regardless of the array size.

Sample Input:
Sorted Array: {2, 5, 7, 10, 12, 15, 18}
Key: 12

Sample Output:
Key 12 found at index 4
*/

#include <iostream>

// Function to perform binary search
int binarySearch(int arr[], int size, int key) {
    int low = 0;
    int high = size - 1;

    while (low <= high) {
        int mid = low + (high - low) / 2;

        if (arr[mid] == key) {
            // If key is found, return its index
            return mid;
        } else if (arr[mid] < key) {
            // If key is greater, search the right half of the array
            low = mid + 1;
        } else {
            // If key is smaller, search the left half of the array
            high = mid - 1;
        }
    }

    // If key is not found, return -1
    return -1;
}

int main() {
    // Sample input (sorted array)
    int arr[] = {2, 5, 7, 10, 12, 15, 18};
    int size = sizeof(arr) / sizeof(arr[0]);
    int key = 12;

    // Perform binary search to find the key in the array
    int index = binarySearch(arr, size, key);

    // Check if the key was found and print the result
    if (index != -1) {
        std::cout << "Key " << key << " found at index " << index << std::endl;
    } else {
        std::cout << "Key " << key << " not found in the array." << std::endl;
    }

    return 0;
}

